Python 3.12 raises deprecation warning related to `datetime.utcnow()`

Using `datetime.utcnow()` triggers a deprecation warning in Python 3.12. This leads to test failures in projects using PyOpenSSL (e.g. [josepy](https://github.com/certbot/josepy)).

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <OpenSSL.crypto.X509 object at 0x7f57096df2c0>

    def has_expired(self) -> bool:
        """
        Check whether the certificate has expired.
    
        :return: ``True`` if the certificate has expired, ``False`` otherwise.
        :rtype: bool
        """
        time_bytes = self.get_notAfter()
        if time_bytes is None:
            raise ValueError("Unable to determine notAfter")
        time_string = time_bytes.decode("utf-8")
        not_after = datetime.datetime.strptime(time_string, "%Y%m%d%H%M%SZ")
    
>       return not_after < datetime.datetime.utcnow()
E       DeprecationWarning: datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.now(datetime.UTC).

/usr/lib/python3.12/site-packages/OpenSSL/crypto.py:1437: DeprecationWarning
=========================== short test summary info ============================
FAILED tests/util_test.py::ComparableX509Test::test_getattr_proxy - Deprecati...
======================== 1 failed, 193 passed in 0.51s =========================
```

How fun. Is the suggested solution something that works on Python 3.7?

On Thu, Jul 6, 2023, 4:41 PM Felix Schwarz ***@***.***> wrote:

> Using datetime.utcnow() triggers a deprecation warning in Python 3.12.
> This leads to test failures in projects using PyOpenSSL (e.g. josepy
> <https://github.com/certbot/josepy>).
>
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
>
> self = <OpenSSL.crypto.X509 object at 0x7f57096df2c0>
>
>     def has_expired(self) -> bool:
>         """
>         Check whether the certificate has expired.
>
>         :return: ``True`` if the certificate has expired, ``False`` otherwise.
>         :rtype: bool
>         """
>         time_bytes = self.get_notAfter()
>         if time_bytes is None:
>             raise ValueError("Unable to determine notAfter")
>         time_string = time_bytes.decode("utf-8")
>         not_after = datetime.datetime.strptime(time_string, "%Y%m%d%H%M%SZ")
>
> >       return not_after < datetime.datetime.utcnow()
> E       DeprecationWarning: datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.now(datetime.UTC).
>
> /usr/lib/python3.12/site-packages/OpenSSL/crypto.py:1437: DeprecationWarning
> =========================== short test summary info ============================
> FAILED tests/util_test.py::ComparableX509Test::test_getattr_proxy - Deprecati...
> ======================== 1 failed, 193 passed in 0.51s =========================
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pyca/pyopenssl/issues/1228>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAGBHWMLZRKWJ7FNZWW6DXO4PIHANCNFSM6AAAAAA2A6ATNM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/src/OpenSSL/crypto.py
+++ b/src/OpenSSL/crypto.py
@@ -1437,7 +1437,9 @@ class X509:
         time_string = time_bytes.decode("utf-8")
         not_after = datetime.datetime.strptime(time_string, "%Y%m%d%H%M%SZ")
 
-        return not_after < datetime.datetime.utcnow()
+        UTC = datetime.timezone.utc
+        utcnow = datetime.datetime.now(UTC).replace(tzinfo=None)
+        return not_after < utcnow
 
     def _get_boundary_time(self, which: Any) -> Optional[bytes]:
         return _get_asn1_time(which(self._x509))

--- a/tests/test_crypto.py
+++ b/tests/test_crypto.py
@@ -7,7 +7,7 @@ Unit tests for :py:mod:`OpenSSL.crypto`.
 import base64
 import sys
 import warnings
-from datetime import datetime, timedelta
+from datetime import datetime, timedelta, timezone
 from subprocess import PIPE, Popen
 
 from cryptography import x509
@@ -66,6 +66,10 @@ from .util import (
 
 def normalize_privatekey_pem(pem):
     return dump_privatekey(FILETYPE_PEM, load_privatekey(FILETYPE_PEM, pem))
+
+
+def utcnow():
+    return datetime.now(timezone.utc).replace(tzinfo=None)
 
 
 GOOD_CIPHER = "blowfish"
@@ -1865,14 +1869,14 @@ class TestX509(_PKeyInteractionTestsMixin):
         current time plus the number of seconds passed in.
         """
         cert = load_certificate(FILETYPE_PEM, self.pemData)
-        not_before_min = datetime.utcnow().replace(microsecond=0) + timedelta(
+        not_before_min = utcnow().replace(microsecond=0) + timedelta(
             seconds=100
         )
         cert.gmtime_adj_notBefore(100)
         not_before = datetime.strptime(
             cert.get_notBefore().decode(), "%Y%m%d%H%M%SZ"
         )
-        not_before_max = datetime.utcnow() + timedelta(seconds=100)
+        not_before_max = utcnow() + timedelta(seconds=100)
         assert not_before_min <= not_before <= not_before_max
 
     def test_gmtime_adj_notAfter_wrong_args(self):
@@ -1891,14 +1895,14 @@ class TestX509(_PKeyInteractionTestsMixin):
         to be the current time plus the number of seconds passed in.
         """
         cert = load_certificate(FILETYPE_PEM, self.pemData)
-        not_after_min = datetime.utcnow().replace(microsecond=0) + timedelta(
+        not_after_min = utcnow().replace(microsecond=0) + timedelta(
             seconds=100
         )
         cert.gmtime_adj_notAfter(100)
         not_after = datetime.strptime(
             cert.get_notAfter().decode(), "%Y%m%d%H%M%SZ"
         )
-        not_after_max = datetime.utcnow() + timedelta(seconds=100)
+        not_after_max = utcnow() + timedelta(seconds=100)
         assert not_after_min <= not_after <= not_after_max
 
     def test_has_expired(self):

